Add blog post for introducing new documentation with OpenAPI and Swagger UI (#278)
openSUSE/obs-landing: _posts/development/2021-01-28-api-documentation-remake.md

---
layout: post
title: API Documentation Remake
category: development
---

> The OBS API has grown over more than fifteen years, and updating and improving the documentation provided to OBS API users has

Reviewer: ```suggestion
The OBS API has grown for more than fifteen years, and updating and improving the documentation provided to OBS API users has
```

become a recurrent request.

Adapting and creating clear OBS documentation is one of our main tasks.
Moved by the success experienced in other teams at [SUSE](https://www.suse.com/), we studied the option of both, creating an
OpenAPI definition, and presenting it through Swagger UI.
After giving it a try with a minimal subset of our API endpoints, we decided it was worth it to arrange an OpenAPI definition.

> So here you have: in the last days we documented, using OpenAPI, a list of endpoints, which will grow in the next iterations.

Reviewer 2: ```suggestion
In the last days we documented, using OpenAPI, a list of endpoints, which will grow in the next iterations. So here you have:
```

Reviewer 2: What do you think?

Author: It's better what you propose. Changed.

<figure>
  <img src="/images/posts/sprint_91_swagger_ui.png" alt="OBS API New Documentation Page" />
  <figcaption>OBS API New Documentation Page</figcaption>
</figure>

> This documentation page is now available in each OBS intance, from the master GitHub branch.

Reviewer 2: ```suggestion
This documentation page is now available in each OBS instance, from the master GitHub branch.
```

> This documentation page is now available in each OBS instance, from the master GitHub branch.

Reviewer: Not sure if you need to mention the whole master branch story, don't remember we were talking about this in other feature blog post, the link is enough in my opinion

Author: Agree. I removed the line. Please have a look again.

For our reference server, you can visit it [here](https://build.opensuse.org/apidocs-new/).

Something to notice, is that the new documentation is not defining schemas to be used for XML validations, for requests and responses.
It will refer to the existing XML validation schemas, also found in the still usable [API documentation](https://build.opensuse.org/apidocs/).

If you are familiar with the [OpenAPI specification](https://www.openapis.org/), you already should know that it is "a specification for
machine-readable interface files for describing, producing, consuming, and visualizing RESTful web services" [1].
Once an OpenAPI definition based in the OpenAPI specification is created, a complete ecosystem of tools can be used.
One of this tools is Swagger UI, that provides a web interface for the documentation included in the OpenAPI definition.

If you are also familiar with [Swagger UI](https://swagger.io/tools/swagger-ui/), you should know that it not only provides an

> ordered list of the API endpoints, grouped, colored by action, examples of parameters and responses, ...

Reviewer 2: ```suggestion
ordered list of the API endpoints, grouped, color-coded by action, examples of parameters and responses, ...
```

Reviewer 2: How about that?

Author: Again, better. Applied.

It also allows you to test your API directly from your browser: asks for authentication when required, asks for the parameters needed,
and shows the results of the performed request.

> As always, we kindly ask you to navigate through this new API documentation, try it, and let us know how was your new experience.

Reviewer: ```suggestion
As always, we kindly ask you to navigate through the new API documentation, try it, and let us know about your experience.
```

---
layout: post
title: API Documentation Remake
category: development
---

The OBS API has grown over more than fifteen years, and updating and improving the documentation provided to OBS API users has
become a recurrent request.

Adapting and creating clear OBS documentation is one of our main tasks.
Moved by the success experienced in other teams at [SUSE](https://www.suse.com/), we studied the option of both, creating an
OpenAPI definition, and presenting it through Swagger UI.
After giving it a try with a minimal subset of our API endpoints, we decided it was worth it to arrange an OpenAPI definition.

In the last days we documented, using OpenAPI, a list of endpoints, which will grow in the next iterations. So here you have:

<figure>
  <img src="/images/posts/sprint_91_swagger_ui.png" alt="OBS API New Documentation Page" />
  <figcaption>OBS API New Documentation Page</figcaption>
</figure>

This documentation page is now available in each OBS instance, from the master GitHub branch.
For our reference server, you can visit it [here](https://build.opensuse.org/apidocs-new/).

> Something to notice, is that the new documentation is not defining schemas to be used for XML validations, for requests and responses.

Reviewer 3: ```suggestion
Something to notice is that the new documentation is not defining schemas to be used for XML validations, for requests, and responses.
```

Author: I wanted to say here `... to be used for both, requests' XML validations and responses' XML validations`.

Your suggestion `... to be used for XML validations, for requests, and responses` sounds to me like three elements.

Maybe only adding a `both`, like this `... to be used for XML validations, both for requests and responses`?

Any other way to write it clearer, not so long?

Reviewer 3: Yeah, your last suggestion sounds much better :+1:

Author: Changed. Please have a look.

It will refer to the existing XML validation schemas, also found in the still usable [API documentation](https://build.opensuse.org/apidocs/).

> If you are familiar with the [OpenAPI specification](https://www.openapis.org/), you already should know that it is "a specification for

Reviewer 3: Nitpick: Instead of maybe writing it like `if you are familiar with Open API/ Swagger UI, it might be better to simply state one or 2 main advantages of them.

Author: I removed the 'If you are familiar with ...' in both sections. Please have a look again.

machine-readable interface files for describing, producing, consuming, and visualizing RESTful web services" (according to
[Wikipedia](https://en.wikipedia.org/wiki/OpenAPI_Specification)).

> Once an OpenAPI definition based in the OpenAPI specification is created, a complete ecosystem of tools can be used.

Reviewer 3: ```suggestion
Once a definition based on the OpenAPI specification is created, a complete ecosystem of tools can be used.
```

> One of this tools is Swagger UI, that provides a web interface for the documentation included in the OpenAPI definition.

Reviewer 3: ```suggestion
One of these tools is Swagger UI, which provides a web interface for the documentation included in the OpenAPI definition.
```

If you are also familiar with [Swagger UI](https://swagger.io/tools/swagger-ui/), you should know that it not only provides an
ordered list of the API endpoints, grouped, color-coded by action, examples of parameters and responses, ...
It also allows you to test your API directly from your browser: asks for authentication when required, asks for the parameters needed,
and shows the results of the performed request.

> As always, we kindly ask you to navigate through this new page, try it, and tell us how was your new experience.

Reviewer 3: ```suggestion
As always, we kindly ask you to navigate through this new API documentation, try it, and let us know how was your new experience.
```

> As always, we kindly ask you to navigate through this new page, try it, and tell us how was your new experience.

Reviewer 4: ```suggestion
As always, we kindly ask you to navigate through this new documentation, try it, and tell us how your new experience was.
```

Author: Applying Abid suggestion made yours outdated.
